### DISCUSSION TITLE
Parent/child component semantics

### DISCUSSION THREAD

**Discussion by denchic45** [evant/kotlin-inject#230] [General]
First I want to say thank you for your library, it is really very useful!

I was a little confused by the section on component arguments where the child component contains the parent component. It seems to me that here it is necessary to consider the child, on the contrary, as a parent who knows about his children, who give their dependencies, or else be considered as just other dependent components. Otherwise, it turns out that the child component can have an unlimited number of parents.
```
@Component
abstract class ChildComponent(
    @Component val parent1Component: Parent1Component,
    @Component val parent2Component: Parent2Component,
    @Component val parent3Component: Parent3Component
) {
    abstract val repo1: Repo1
    abstract val repo2: Repo2
    abstract val repo3: Repo3
}

@Component
abstract class Parent1Component {
    @Provides
    fun provideParent1() = Repo1()
}

@Component
abstract class Parent2Component {
    @Provides
    fun provideParent2() = Repo2()
}

@Component
abstract class Parent3Component {
    @Provides
    fun provideParent3() = Repo3()
}

@Inject
class Repo1

@Inject
class Repo2

@Inject
class Repo3


   fun main() {
  ChildComponent::class.create(
            Parent1Component::class.create(),
            Parent2Component::class.create(),
            Parent3Component::class.create()
        )
    }
```

**Comment by evant** [comment]
Interesting point. I was thinking about it in perspective of lifetime, i.e the parent lives longer than the child. I don't really have strong opinions on the terminology here and would be open to alternatives. For reference, they are pretty close to component dependencies in dagger and I wouldn't be against calling them that.
